File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.cpp
class Solution {
public:
    vector<int> getFinalState(vector<int>& nums, int k, int multiplier) {
        int MOD = 1000000007;
        
        int n = nums.size();
        
        map<int, set<int>> mp;
        for(int i = 0; i < n; i++) {
            mp[nums[i]].insert(i);
        }
        
        while(k-- && mp.size() > 0) {
            long minNum = mp.begin()->first;
            int minNumIndex = *(mp[minNum].begin());
            mp[minNum].erase(minNumIndex);
            if(mp[minNum].size() == 0) {
                mp.erase(minNum);
            }
            int newNum = (minNum * (long)multiplier) % MOD;
            mp[newNum].insert(minNumIndex);
        }
        
        for(auto& data : mp) {
            int num = data.first;
            for(int i : data.second) {
                nums[i] = num;
            }
        }
        
        for(int i = 0; i < n; i++) {
            nums[i] = (nums[i] % MOD);
        }
        
        return nums;
    }
};

File: 3264-final-array-state-after-k-multiplication-operations-i/3264-final-array-state-after-k-multiplication-operations-i.java
class Solution {
    public int[] getFinalState(int[] nums, int k, int multiplier) {
        int n = nums.length;
        
        PriorityQueue<int[]> minHeap = new PriorityQueue<>(new Comparator<int[]>(){
           public int compare(int[] a, int[] b) {
               if(a[0] == b[0]) {
                   return a[1] - b[1];
               }
               return a[0] - b[0];
           } 
        });
        
        for(int i = 0; i < n; i++) {
            minHeap.add(new int[]{nums[i], i});
        }
        
        while(k-- > 0) {
            int[] currNum = minHeap.poll();
            minHeap.add(new int[]{multiplier * currNum[0], currNum[1]});
        }
        
        while(!minHeap.isEmpty()) {
            int[] currNum = minHeap.poll();
            nums[currNum[1]] = currNum[0]; 
        }
        
        return nums;
    }
}